src/Datadog.Trace.ClrProfiler.Native/cor_profiler.cpp

```diff
@@ -597,6 +649,666 @@ HRESULT STDMETHODCALLTYPE CorProfiler::JITCompilationStarted(
+                                          L"System.Runtime.InteropServices.Marshal",
+                                          &marshal_type_ref);
+  if (FAILED(hr)) {
+    Warn("GenerateVoidILStartupMethod: DefineTypeRefByName failed");
+    return hr;
+  }
+
+  // Get a MemberRef for System.Runtime.InteropServices.Marshal.Copy(IntPtr, Byte[], int, int)
+  mdMemberRef marshal_copy_member_ref;
+  COR_SIGNATURE marshal_copy_signature[] = {
+      IMAGE_CEE_CS_CALLCONV_DEFAULT, // Calling convention
+      4,                             // Number of parameters
+      ELEMENT_TYPE_VOID,             // Return type
+      ELEMENT_TYPE_I,                // List of parameter types
+      ELEMENT_TYPE_SZARRAY,
+      ELEMENT_TYPE_U1,
+      ELEMENT_TYPE_I4,
+      ELEMENT_TYPE_I4
+  };
+  hr = metadata_emit->DefineMemberRef(
+      marshal_type_ref, L"Copy", marshal_copy_signature,
+      sizeof(marshal_copy_signature), &marshal_copy_member_ref);
+  if (FAILED(hr)) {
+    Warn("GenerateVoidILStartupMethod: DefineMemberRef failed");
+    return hr;
+  }
+
+  // Get a TypeRef for System.Reflection.Assembly
+  mdTypeRef system_reflection_assembly_type_ref;
+  hr = metadata_emit->DefineTypeRefByName(mscorlib_ref,
+                                          L"System.Reflection.Assembly",
+                                          &system_reflection_assembly_type_ref);
+  if (FAILED(hr)) {
+    Warn("GenerateVoidILStartupMethod: DefineTypeRefByName failed");
+    return hr;
+  }
+
+  // Get a MemberRef for System.Object.ToString()
+  mdTypeRef system_object_type_ref;
+  hr = metadata_emit->DefineTypeRefByName(mscorlib_ref,
+                                          L"System.Object",
+                                          &system_object_type_ref);
+  if (FAILED(hr)) {
+    Warn("GenerateVoidILStartupMethod: DefineTypeRefByName failed");
+    return hr;
+  }
+
+  // Get a TypeRef for System.AppDomain
+  mdTypeRef system_appdomain_type_ref;
+  hr = metadata_emit->DefineTypeRefByName(mscorlib_ref,
+                                          L"System.AppDomain",
+                                          &system_appdomain_type_ref);
+  if (FAILED(hr)) {
+    Warn("GenerateVoidILStartupMethod: DefineTypeRefByName failed");
+    return hr;
+  }
+
+  // Get a MemberRef for System.AppDomain.get_CurrentDomain()
+  // and System.AppDomain.Assembly.Load(byte[], byte[])
+
+  // Create method signature for AppDomain.CurrentDomain property
+  BYTE appdomain_get_current_domain_signature[7];
+  COR_SIGNATURE appdomain_get_current_domain_signature_start[] = {
+      IMAGE_CEE_CS_CALLCONV_DEFAULT,
+      0,
+      ELEMENT_TYPE_CLASS, // ret = System.AppDomain
+      // insert compressed token for System.AppDomain TypeRef here
+  };
+
+  memcpy(appdomain_get_current_domain_signature,
+         appdomain_get_current_domain_signature_start,
+         sizeof(appdomain_get_current_domain_signature_start));
+  CorSigCompressToken(system_appdomain_type_ref,
```

Comment:
I think you need to look at the return value from `CorSigCompressToken()` to know the size of the compressed token. Otherwise, you don't how many of the bytes in `appdomain_get_current_domain_signature` have valid data. You will won't always use all 7 bytes and will have a few "left over" bytes at the end.

Reply:
For this point, I can certainly make the size exact by dynamically allocating a new array. Though, having extra space in the array doesn't cause issues because the signature token specifies how many parameters to expect so the parsing is still correct.

Reply:
I'll follow up with the suggested change.

Reply:
Addressed in the latest commit, let me know if that looks good!